go/vt/topo/topo_lock.go

```diff
@@ -0,0 +1,169 @@
+}
+
+// unlock unlocks a previously locked key.
+func (l *Lock) unlock(ctx context.Context, path string, lockDescriptor LockDescriptor, actionError error) error {
+	// Detach from the parent timeout, but copy the trace span.
+	// We need to still release the lock even if the parent
+	// context timed out.
+	ctx = trace.CopySpan(context.TODO(), ctx)
+	ctx, cancel := context.WithTimeout(ctx, RemoteOperationTimeout)
+	defer cancel()
+
+	span, ctx := trace.NewSpan(ctx, "TopoServer.Unlock")
+	span.Annotate("action", l.Action)
+	span.Annotate("path", path)
+	defer span.Finish()
+
+	// first update the actionNode
+	if actionError != nil {
+		l.Status = "Error: " + actionError.Error()
+	} else {
+		l.Status = "Done"
+	}
+	return lockDescriptor.Unlock(ctx)
+}
+
+// Lock adds lock information to the context, checks that the lock is not already held, and locks it.
+// It returns a new context with the lock information and a function to unlock the lock.
+func (tl TopoLock) Lock(ctx context.Context) (context.Context, func(*error), error) {
+	i, ok := ctx.Value(locksKey).(*locksInfo)
+	if !ok {
+		i = &locksInfo{
+			info: make(map[string]*lockInfo),
+		}
+		ctx = context.WithValue(ctx, locksKey, i)
+	}
+	i.mu.Lock()
+	defer i.mu.Unlock()
+	// check that we are not already locked
+	if _, ok := i.info[tl.Path]; ok {
+		return nil, nil, vterrors.Errorf(vtrpc.Code_INVALID_ARGUMENT, "lock for %v is already held", tl.Path)
+	}
+
+	// lock it
+	l := newLock(fmt.Sprintf("lock for %s", tl.Name))
+	lockDescriptor, err := l.lock(ctx, tl.ts, tl.Path)
+	if err != nil {
+		return nil, nil, err
+	}
+	// and update our structure
+	i.info[tl.Path] = &lockInfo{
+		lockDescriptor: lockDescriptor,
+		actionNode:     l,
+	}
+	return ctx, func(finalErr *error) {
+		i.mu.Lock()
+		defer i.mu.Unlock()
+
+		if _, ok := i.info[tl.Path]; !ok {
+			if *finalErr != nil {
+				log.Errorf("trying to unlock %v multiple times", tl.Path)
+			} else {
+				*finalErr = vterrors.Errorf(vtrpc.Code_INVALID_ARGUMENT, "trying to unlock %v multiple times", tl.Path)
+			}
+			return
+		}
+
+		err := l.unlock(ctx, tl.Path, lockDescriptor, *finalErr)
+		// if we have an error, we log it, but we still want to delete the lock
+		if *finalErr != nil {
+			if err != nil {
+				// both error are set, just log the unlock error
+				log.Errorf("unlock(%v) failed: %v", tl.Path, err)
+			}
+		} else {
+			*finalErr = err
+		}
+		delete(i.info, tl.Path)
+	}, nil
+}
+
+func CheckLocked(ctx context.Context, keyPath string) error {
+	// extract the locksInfo pointer
+	i, ok := ctx.Value(locksKey).(*locksInfo)
+	if !ok {
+		return vterrors.Errorf(vtrpc.Code_INVALID_ARGUMENT, "%s is not locked (no locksInfo)", keyPath)
+	}
+	i.mu.Lock()
+	defer i.mu.Unlock()
+
+	// find the individual entry
+	_, ok = i.info[keyPath]
+	if !ok {
+		return vterrors.Errorf(vtrpc.Code_INVALID_ARGUMENT, "%s is not locked (no lockInfo in map)", keyPath)
+	}
+
+	// and we're good for now.
+	return nil
```

Comment:
I think this might be an insufficient check for knowing if the lock still exists. Because etcd expires the lock after a certain amount of time, I think we need something like - 
```go
// find the individual entry
	li, ok := i.info[keyPath]
	if !ok {
		return vterrors.Errorf(vtrpc.Code_INVALID_ARGUMENT, "%s is not locked (no lockInfo in map)", keyPath)
	}

	// check if we still have the lock
	return li.lockDescriptor.Check(ctx)
```

I added some docs on how checking of shard lock works for different topo servers, but I think we still need to do this. This is what we are doing in `CheckShardLocked` as well.

Reply:
Right. We need to add it to `CheckKeyspaceLocked`  too ...